src/ripple/shamap/README.md

```diff
@@ -26,164 +56,292 @@ There are two different ways of building and using a SHAMap:
+some piece of code wishes to query the state of the ledger.  In this case we
+don't wish to change the state of the SHAMap, so we'd use an immutable snapshot.
+
+## Sequence numbers ##
+
+Both SHAMap's and their nodes carry a sequence number.  This is simply an
+unsigned number that indicates ownership or membership, or a non-membership.
 
-An example using a immutable ledger would be when there's an open ledger
-and some piece of code wishes to query the state of the ledger.  In this
-case we don't wish to change the state of the SHAMap, so we'd use an
-immutable snapshot.
+SHAMap's sequence numbers normally start out as 1.  However when a snap-shot of
+a SHAMap is made, the copy's sequence number is 1 greater than the original.
 
+The nodes of a SHAMap have their own copy of a sequence number.  If the SHAMap
+is not immutable, meaning it can change, then all of its nodes must have the
+same sequence number as the SHAMap itself.  This enforces an invariant that none
+of the nodes are shared with other SHAMaps.
+
+When a SHAMap needs to have a private copy of a node, not shared by any other
+SHAMap, it first clones it and then sets the new copy to have a sequence number
+equal to the SHAMap sequence number.  The `unshareNode` is a private utility
+which automates the task of first checking if the node is already sharable, and
+if so, cloning it and giving it the proper sequence number.  An example case
+where a private copy is needed is when an inner node needs to have a child
+pointer altered.  Any modification to a node will require a non-shared node.
+
+When a SHAMap decides that it is safe to share a node of its own, it sets the
+node's sequence number to 0 (a SHAMap never has a sequence number of 0). This
+is done for every node in the trie when `SHAMap::walkSubTree` is executed.
+
+Note that other objects in rippled also have sequence numbers (e.g. ledgers).
+The SHAMap and node sequence numbers should not be confused with these other
+sequence numbers (no relation).
 
 ## SHAMap Creation ##
 
 A SHAMap is usually not created from vacuum.  Once an initial SHAMap is
-constructed, later SHAMaps are usually created by calling
-snapShot(bool isMutable) on the original SHAMap().  The returned SHAMap
-has the expected characteristics (mutable or immutable) based on the passed
-in flag.
+constructed, later SHAMaps are usually created by calling snapShot(bool
+isMutable) on the original SHAMap().  The returned SHAMap has the expected
+characteristics (mutable or immutable) based on the passed in flag.
 
 It is cheaper to make an immutable snapshot of a SHAMap than to make a mutable
-snapshot.  If the SHAMap snapshot is mutable then any of the nodes that might
-be modified must be copied before they are placed in the mutable map.
+snapshot.  If the SHAMap snapshot is mutable then any of the nodes that might be
+modified must be copied before they are placed in the mutable map.
```

Comment:
I see later that there is mention of copy-on-write, but I'm not sure why it doesn't happen here.

Reply:
I've clarified the wording to say that it does happen here.